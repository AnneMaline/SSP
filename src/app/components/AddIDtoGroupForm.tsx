"use client";
import { useState } from "react";

type AddIDtoGroupFormType = {
  entraID: string;
  environment: "prod" | "test" | "development";
  group: string;
};

const AddIDtoGroupForm = () => {
  // Initial state for the form
  const initialFormData: AddIDtoGroupFormType = {
    entraID: "",
    environment: "prod",
    group: "",
  };
  const environments: AddIDtoGroupFormType["environment"][] = [
    "prod",
    "test",
    "development",
  ];

  // Dummy dynamic groups (update when the API is connected)
  const [groups, setGroups] = useState<string[]>([
    "Group 1",
    "Group 2",
    "Group 3",
  ]);

  // feedback and questions form
  const [formData, setFormData] =
    useState<AddIDtoGroupFormType>(initialFormData);

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    console.log("Submitting form data:", formData);

    // Reset form after submission
    setFormData(initialFormData);

    /*// Send data to the server or API
    const response = await fetch("/api/submit-form", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(formData),
    });

    if (response.ok) {
      console.log("Form submitted successfully!");
    } else {
      console.error("Form submission failed.");
    }*/
  };
  return (
    <form
      onSubmit={handleSubmit}
      className="max-w-md mx-auto p-6 border border-gray-300 rounded-lg shadow space-y-6"
    >
      {/* EntraID */}
      <div className="space-y-2">
        <label htmlFor="entraID" className="block text-lg font-semibold">
          Enter ID
        </label>
        <input
          required
          type="text"
          id="entraID"
          name="entraID"
          value={formData.entraID}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setFormData((prev) => ({
              ...prev,
              entraID: e.target.value,
            }))
          }
          className="w-full px-3 py-2 border border-gray-300 rounded"
          placeholder="Enter entraID"
        />
      </div>

      {/* Environment */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Environment</legend>
        <div className="space-y-1">
          {environments.map((type) => (
            <label key={type} className="block">
              <input
                type="radio"
                name="environment"
                value={type}
                checked={formData.environment === type}
                onChange={() =>
                  setFormData((prev) => ({
                    ...prev,
                    environment: type as AddIDtoGroupFormType["environment"],
                  }))
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Group */}
      <div className="space-y-2">
        <label htmlFor="group" className="block text-lg font-semibold">
          Select Group
        </label>
        <select
          required
          id="group"
          name="group"
          value={formData.group}
          onChange={(e: React.ChangeEvent<HTMLSelectElement>) =>
            setFormData((prev) => ({
              ...prev,
              group: e.target.value,
            }))
          }
          className="w-full px-3 py-2 border border-gray-300 rounded"
        >
          <option value="">Select a group</option>
          {groups.map((group, index) => (
            <option key={index} value={group}>
              {group}
            </option>
          ))}
        </select>
      </div>

      {/* Submit Button */}
      <button
        type="submit"
        className="w-full bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-600"
      >
        Add
      </button>
    </form>
  );
};

export default AddIDtoGroupForm;
