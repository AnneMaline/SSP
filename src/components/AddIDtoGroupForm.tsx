"use client";
import { addMember } from "@/utils/entitlement/addMember";
import { getGroups } from "@/utils/entitlement/getGroups";
import { useSession } from "next-auth/react";
import { useEffect, useState } from "react";

type Environment = "prod" | "test" | "development" | "bootcamp";

type AddIDtoGroupFormType = {
  entraID: string;
  environment: Environment[];
  role: "MEMBER" | "OWNER";
  group: string;
};

type Groups = {
  name: string;
  email: string;
  description: string;
};

type AddIDtoGroupFormProps = {
  group?: string;
};

const AddIDtoGroupForm = ({ group }: AddIDtoGroupFormProps) => {
  // Initial state for the form
  const initialFormData: AddIDtoGroupFormType = {
    entraID: "",
    environment: [],
    role: "MEMBER",
    group: group || "",
  };
  const environments: AddIDtoGroupFormType["environment"] = [
    "prod",
    "test",
    "development",
    "bootcamp",
  ];

  // ----------------------Fetch groups----------------------
  const [groups, setGroups] = useState<Groups[]>([]);
  const data_partition_id = "bootcamp";
  const { data: session } = useSession();

  useEffect(() => {
    if (!session || !session.accessToken) {
      throw new Error("Session not found");
    }
    getGroups(data_partition_id, session.accessToken).then((groups) => {
      setGroups(groups);
      //setFormData((prev) => ({ ...prev, group: groups })); //TASK: Make this work. Initial group in the form
    });
  }, [session]);

  // ----------------------Make and submitt form----------------------
  const [formData, setFormData] =
    useState<AddIDtoGroupFormType>(initialFormData);
  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    // prevent the default form submission
    e.preventDefault();

    // if environment is empty
    if (formData.environment.length === 0) {
      alert("Please select at least one environment");
      return;
    }

    // Reset form after submission
    setFormData(initialFormData);

    // Send data to the API
    if (session && session.accessToken) {
      if (group) {
        addMember(
          formData.entraID,
          formData.role,
          "bootcamp",
          group,
          session.accessToken
        );
      } else {
        addMember(
          formData.entraID,
          formData.role,
          "bootcamp",
          formData.group,
          session.accessToken
        );
      }
    }
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="max-w-md mx-auto p-6 border-2 border-dashed border-[#80AFB4] rounded-lg space-y-6"
    >
      {/* EntraID */}
      <div className="space-y-2">
        <label htmlFor="entraID" className="block text-lg font-semibold">
          Enter ID
        </label>
        <input
          required
          type="text"
          id="entraID"
          name="entraID"
          value={formData.entraID}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setFormData((prev) => ({
              ...prev,
              entraID: e.target.value,
            }))
          }
          className="w-full px-3 py-2 border border-gray-300 rounded"
          placeholder="Enter entraID"
        />
      </div>

      {/* Role */}
      <div className="space-y-2">
        <legend className="block text-lg font-semibold">Select Role</legend>
        <div className="space-y-1">
          <label className="block">
            <input
              type="radio"
              name="role"
              value="MEMBER"
              checked={formData.role === "MEMBER"}
              onChange={() =>
                setFormData((prev) => ({
                  ...prev,
                  role: "MEMBER",
                }))
              }
              className="mr-2"
            />
            Member
          </label>
          <label className="block">
            <input
              type="radio"
              name="role"
              value="OWNER"
              checked={formData.role === "OWNER"}
              onChange={() =>
                setFormData((prev) => ({
                  ...prev,
                  role: "OWNER",
                }))
              }
              className="mr-2"
            />
            Owner
          </label>
        </div>
      </div>

      {/* Environment */}
      <fieldset className="space-y-2">
        <legend className="block text-lg font-semibold">Environment</legend>
        <div className="space-y-1">
          {environments.map((type) => (
            <label key={type} className="block">
              <input
                type="checkbox"
                name="environment"
                value={type}
                checked={formData.environment.includes(type)}
                onChange={() =>
                  setFormData((prev) => {
                    const newEnvironments = prev.environment.includes(type)
                      ? prev.environment.filter((env) => env !== type)
                      : [...prev.environment, type];
                    return {
                      ...prev,
                      environment: newEnvironments,
                    };
                  })
                }
                className="mr-2"
              />
              {type.charAt(0).toUpperCase() + type.slice(1)}
            </label>
          ))}
        </div>
      </fieldset>

      {/* Group */}
      <div className="space-y-2">
        <label htmlFor="group" className="block text-lg font-semibold">
          {group ? "Selected Group" : "Select Group"}
        </label>
        {group ? (
          <div>{group}</div>
        ) : (
          <select
            required
            id="group"
            name="group"
            value={formData.group}
            onChange={(e: React.ChangeEvent<HTMLSelectElement>) =>
              setFormData((prev) => ({
                ...prev,
                group: e.target.value,
              }))
            }
            className="w-full px-3 py-2 border border-gray-300 rounded"
          >
            <option value="">Select a group</option>
            {groups.map((group) => (
              <option key={group.email} value={group.email}>
                {group.name}
              </option>
            ))}
          </select>
        )}
      </div>

      {/* Submit Button */}
      <button
        type="submit"
        className="w-full bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-600"
      >
        Add
      </button>
    </form>
  );
};

export default AddIDtoGroupForm;
