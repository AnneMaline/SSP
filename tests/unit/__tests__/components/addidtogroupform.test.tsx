import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import AddIDtoGroupForm from "../../../../src/components/AddIDtoGroupForm";

describe("AddIDtoGroupForm Component", () => {
  it("renders correctly", () => {
    const { container } = render(<AddIDtoGroupForm />);
    expect(container).toMatchSnapshot();
  });

  it("renders the form with all fields", () => {
    render(<AddIDtoGroupForm />);

    // Check EntraID input
    expect(screen.getByLabelText("Enter ID")).toBeInTheDocument();

    // Check environment radio buttons
    expect(screen.getByLabelText("Prod")).toBeInTheDocument();
    expect(screen.getByLabelText("Test")).toBeInTheDocument();
    expect(screen.getByLabelText("Development")).toBeInTheDocument();

    // Check group dropdown
    expect(screen.getByLabelText("Select Group")).toBeInTheDocument();

    // Check submit button
    expect(screen.getByRole("button", { name: "Add" })).toBeInTheDocument();
  });

  it("updates the EntraID field when typed", () => {
    render(<AddIDtoGroupForm />);
    const entraIDInput = screen.getByLabelText("Enter ID");

    // Check initial state
    expect(entraIDInput).toHaveValue("");

    // Type into the EntraID input
    fireEvent.change(entraIDInput, { target: { value: "12345" } });

    // Verify that the input value was updated
    expect(entraIDInput).toHaveValue("12345");
  });

  it("updates the environment when a radio button is selected", () => {
    render(<AddIDtoGroupForm />);
    const prodRadio = screen.getByLabelText("Prod");
    const testRadio = screen.getByLabelText("Test");

    // Check initial state (default should be prod)
    expect(prodRadio).toBeChecked();
    expect(testRadio).not.toBeChecked();

    // Select "Test"
    fireEvent.click(testRadio);

    // Verify that the "Test" radio is selected and "Prod" is deselected
    expect(testRadio).toBeChecked();
    expect(prodRadio).not.toBeChecked();
  });

  it("updates the group value when a dropdown option is selected", async () => {
    render(<AddIDtoGroupForm />);

    // Wait for groups to load (simulated async loading)
    await waitFor(() =>
      expect(screen.getByLabelText("Select Group")).toBeInTheDocument()
    );

    const groupSelect = screen.getByLabelText("Select Group");

    // Check initial state (should be empty)
    expect(groupSelect).toHaveValue("");

    // Select a group
    fireEvent.change(groupSelect, { target: { value: "Group 1" } });

    // Verify that the selected group is updated
    expect(groupSelect).toHaveValue("Group 1");
  });

  it("resets the form after submission", async () => {
    render(<AddIDtoGroupForm />);
    const entraIDInput = screen.getByLabelText("Enter ID");
    const groupSelect = screen.getByLabelText("Select Group");
    const submitButton = screen.getByRole("button", { name: "Add" });

    // Fill out the form
    fireEvent.change(entraIDInput, { target: { value: "12345" } });
    fireEvent.change(groupSelect, { target: { value: "Group 1" } });
    fireEvent.click(screen.getByLabelText("Development"));

    // Submit the form
    fireEvent.click(submitButton);

    // Verify that the form is reset after submission
    await waitFor(() => expect(entraIDInput).toHaveValue(""));
    await waitFor(() => expect(groupSelect).toHaveValue(""));
    await waitFor(() => expect(screen.getByLabelText("Prod")).toBeChecked());
  });
});
