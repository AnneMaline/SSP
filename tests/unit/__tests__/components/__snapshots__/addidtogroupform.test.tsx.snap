// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddIDtoGroupForm Component renders correctly 1`] = `
<div>
  <form
    class="max-w-md mx-auto p-6 border-2 border-dashed border-[#80AFB4] rounded-lg space-y-6"
  >
    <div
      class="space-y-2"
    >
      <label
        class="block text-lg font-semibold"
        for="entraID"
      >
        Enter ID
      </label>
      <input
        class="w-full px-3 py-2 border border-gray-300 rounded"
        id="entraID"
        name="entraID"
        placeholder="Enter entraID"
        required=""
        type="text"
        value=""
      />
    </div>
    <fieldset
      class="space-y-2"
    >
      <legend
        class="block text-lg font-semibold"
      >
        Environment
      </legend>
      <div
        class="space-y-1"
      >
        <label
          class="block"
        >
          <input
            checked=""
            class="mr-2"
            name="environment"
            type="radio"
            value="prod"
          />
          Prod
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="environment"
            type="radio"
            value="test"
          />
          Test
        </label>
        <label
          class="block"
        >
          <input
            class="mr-2"
            name="environment"
            type="radio"
            value="development"
          />
          Development
        </label>
      </div>
    </fieldset>
    <div
      class="space-y-2"
    >
      <label
        class="block text-lg font-semibold"
        for="group"
      >
        Select Group
      </label>
      <select
        class="w-full px-3 py-2 border border-gray-300 rounded"
        id="group"
        name="group"
        required=""
      >
        <option
          value=""
        >
          Select a group
        </option>
        <option
          value="Group 1"
        >
          Group 1
        </option>
        <option
          value="Group 2"
        >
          Group 2
        </option>
        <option
          value="Group 3"
        >
          Group 3
        </option>
      </select>
    </div>
    <button
      class="w-full bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-600"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
